docs/architecture/adr-040-storage-and-smt-state-commitments.md

```diff
@@ -0,0 +1,118 @@
+# ADR 040: Storage and SMT State Commitments
+
+## Changelog
+
+- 2020-01-15: Draft
+
+## Status
+
+DRAFT Not Implemented
+
+
+## Abstract
+
+Sparse Merke Tree (SMT) is a version of a Merkle Tree with various storage and performance optimizations. This ADR defines a separation of state commitments from data storage and the SDK transition from IAVL to SMT.
```

Comment:
I'm wondering if this shouldn't in fact be two ADRs instead? One for separating storage and commitments and one about the SMT.

Reply:
I was also thinking about it. But they are highly related - one cannot be done without other. Hence, I'm proposing here a general design and leave a space for future ADR for RDMS which will introduce SDK breaking changes.

Reply:
Well we could separate the two with IAVL right? We don't need SMT for that AFAIK...

Reply:
@aaronc, we could describe here only SMT, but it will only a half backed idea without a working solution:
* keeping IAVL (in it's current implementation) with anything else doesn't make sense because we double the data.
* the main value proposition here is to not store objects in SMT (we store only hashes).

Do you have something else in mind?